Mk reconciliation migrations

mkangia opened the pull request:
https://trello.com/c/iedwBirn/204-model-reconciliation-for-public-sector

3 down/ 1 more to go.

mkangia committed:
Add drug resistance reconciliation migration [skip ci]

mkangia committed:
use bulk update instead and add close reason on cases being closed [skip ci]

mkangia committed:
Add multiple open referral cases reconciliation migration [skip ci]

mkangia committed:
lint

mkangia committed:
better naming [skip ci]

mkangia committed:
lint

mkangia committed:
Add multiple valid occurrence and episode cases reconciliation [skip ci]

mkangia committed:
cover reconciliation for episode cases even in case occurrence cases are valid [skip ci]

mkangia committed:
lint [skip ci]

mkangia committed:
rename file to justify what it does[skip ci]

mkangia committed:
flag if either of the condition is true

mkangia committed:
fix logic for closing cases, pass correct set of cases for reconciliation and refactor

mkangia committed:
fix script: pass id where needed, avoid errors due to incorrect way to create lists, fix case close implementation and refactor

mkangia committed:
find only open occurrence cases, pass id when needed, fix case close and refactor [skip ci]

mkangia committed:
lint [skip ci]

sheelio commented:
@mkangia 

These look like they're written like commands, not regular reconcilliation tasks that run.  are these supposed to run nightly?

sheelio commented on custom/enikshay/management/commands/drug_resistance_reconciliation.py:
```diff
@@ -0,0 +1,197 @@
+
+        # group cases by their sensitivity
+        # possible values are resistant/ sensitive/ unknown
+        for drug_resistance_case in drug_resistance_cases:
```
@mkangia 

My idea for how to do this was to to do a multikey sort on the drug_resistance cases
- First sort descending by sensitivity (resistant = 3, sensitive = 2, anything else is 1_
-Then sort by opened_on (ascending)

Then keep the first result.  Not sure if that would be easier to parse / read.

esoergel replied:
You could use `min` and pass a key function which returns a `(preference, opened_on)` tuple (where preference is resistant = 1, sensitive = 2, anything else is 3).  That should do the ordering properly.

esoergel commented on custom/enikshay/management/commands/duplicate_occurrences_and_episodes_reconciliation.py:
```diff
@@ -0,0 +1,228 @@
+        # get all episode cases for all open occurrences
+        all_episode_cases = []
+        for open_occurrence_case_id in open_occurrence_case_ids:
+            all_episode_cases += self.get_open_reconciled_episode_cases_for_occurrence(open_occurrence_case_id)
```
Here you have access to the occurrence cases already, so you could make a `{episode_id: occurrence_case}` map so later on you don't need to look up occurrence cases from the DB.

esoergel commented on custom/enikshay/management/commands/duplicate_occurrences_and_episodes_reconciliation.py:
```diff
@@ -0,0 +1,228 @@
+            raise CommandError("Resolved open active episode cases were not resolved for occurrence, %s" %
+                               occurrence_case_id)
+
+        return _get_open_episode_cases_for_occurrence(occurrence_case_id)
```
You can just return `all_open_episode_cases` here.

mkangia committed:
Use get_case_property instead of getting all properties first, use items for iteration, log in dry_run as well [skip ci]

mkangia committed:
rescue when person case is not found for cases. [skip ci]

These would be the cases from earlier attempts of migration where person cases were removed but others were kept. Safe to skip those.

mkangia committed:
add missing get_id calls and file extension [skip ci]

mkangia committed:
lint [skip ci]

sheelio commented:
@mkangia

These look like they're written like commands, not regular reconcilliation tasks that run. These supposed to run nightly.

mkangia committed:
Add investigations reconciliation

mkangia committed:
Fix script [skip ci]

mkangia committed:
make further progress on investigation reconciliation [skip ci]

mkangia committed:
Add details for extension cases being closed and episode details when needed [skip ci]

mkangia committed:
remove unused variable [skip ci]

mkangia committed:
Get recently modified case by user instead of recently opened to retain [skip ci]

mkangia committed:
correct label [skip ci]

mkangia committed:
Add option [skip ci]

mkangia committed:
get relevant episode case on basis of treatment_card_completed_date date and then by user edited date [skip ci]

mkangia committed:
fix logic for recently used timestamp check [skip ci]

mkangia committed:
Email final report [skip ci]

mkangia committed:
add option to email reports and force dry run on reports that are not yet confirmed

mkangia committed:
add a view to run reconciliation tasks and email the report

mkangia committed:
Add missing header [skip ci]

mkangia committed:
remove unused variable [skip ci]

mkangia commented:
@esoergel Can you review this again. 
Since your last review, just minor improvements/changes to the tasks but mainly added a view for @sheelio or Kriti to run this regularly or whenever needed for diagnosis for reconciliations pending confirmation or real run for confirmed one. This probably will be helpful even when the task is automatically running to run any reconciliation thats needed urgently and can't wait for midnight or whenever the task is to run.

I want to merge this and add follow PR(s) for any changes that would be needed.
Only the Duplicate Episode and Occurrence reconciliation is confirmed so others are forced for dry run as of now.

let me know if you think this can be done better.

mkangia committed:
resolve merge conflict in enikshay/urls.py

mkangia committed:
add person case id to drug resistance reconciliation [skip ci]

mkangia committed:
fix issue when fetching recipient [skip ci]

esoergel commented on custom/enikshay/management/commands/duplicate_occurrences_and_episodes_reconciliation.py:
```diff
@@ -0,0 +1,337 @@
+                recently_modified_time = last_user_edit_on_phone
+                recently_modified_case = case
+            elif recently_modified_time and recently_modified_time == last_user_edit_on_phone:
+                print("This looks like a super edge case that needs to be looked at. Not blocking as of now.")
```
Do you want any identifying info like case IDs here if this does happen?

esoergel commented on custom/enikshay/management/commands/investigations_reconciliation.py:
```diff
@@ -0,0 +1,258 @@
+        )
+        email.attach(filename=self.result_file_name, content=csvfile.read())
+        csvfile.close()
+        email.send()
```
There's a lot of repeated code in these migrations, it would be nice to pull some of that out if possible.

mkangia replied:
Didn't want to extract the code and put in the app so,
adding BaseModelReconciliationCommand and then inherit that.

esoergel commented on custom/enikshay/management/commands/investigations_reconciliation.py:
```diff
@@ -0,0 +1,258 @@
+    open_confirmed_drtb_episode_cases = []
+    for occurrence_case in occurrence_cases:
+        case_accessor = CaseAccessors(DOMAIN)
+        all_cases = case_accessor.get_reverse_indexed_cases([occurrence_case.case_id])
```
I think this could just accept a list of all the occurrence case ids, rather than doing this in a loop and extending the list.

mkangia replied:
Oh right. Updated!

esoergel commented on custom/enikshay/management/commands/investigations_reconciliation.py:
```diff
@@ -0,0 +1,258 @@
+            bulk_update_cases(DOMAIN, updates, self.__module__)
+
+    def get_coalesced_value_for_case_property(self, case_property, all_values):
+        #ToDo: write logic for this
```
There's a ToDo here

mkangia replied:
Ya. Need some cases to verify what should be done here but there are not any of now.

esoergel commented on custom/enikshay/management/commands/investigations_reconciliation.py:
```diff
@@ -0,0 +1,258 @@
+        self.writerow(coalesced_investigation_case)
+
+    def close_or_update_investigation_cases(self, all_cases, retain_case_id, episode_case_id,
+                                            investigation_interval, updates):
```
This method isn't called anywhere

mkangia replied:
This reconciliation is still pending confirmation. So just logging what would happen and not actually making any change here.

Would be called once we want to close or update cases.

esoergel commented on custom/enikshay/management/commands/investigations_reconciliation.py:
```diff
@@ -0,0 +1,258 @@
+        coalesced_investigation_case['modified_on'] = latest_investigation_case.modified_on
+        coalesced_investigation_case['episode_id'] = episode_case_id
+        coalesced_investigation_case['update_or_close'] = 'update'
+        self.writerow(coalesced_investigation_case)
```
Should this also save something to the db?

mkangia replied:
This reconciliation is still pending confirmation. So just logging what would happen and not actually making any change here.

esoergel commented on custom/enikshay/management/commands/investigations_reconciliation.py:
```diff
@@ -0,0 +1,258 @@
+    def close_or_update_investigation_cases(self, all_cases, retain_case_id, episode_case_id,
+                                            investigation_interval, updates):
+        all_case_ids = [investigation_case.case_id for investigation_case in all_cases]
+        # ToDo: refetch investigation_cases in case the len of set is different from list
```
Still actionable?

mkangia replied:
Not worried about this yet. Will look at this when this is needed.

esoergel commented on custom/enikshay/management/commands/multiple_open_referrals_reconciliation.py:
```diff
@@ -0,0 +1,141 @@
+                referral_cases = get_open_referral_cases_from_occurrence(occurrence_case_id)
+                if len(referral_cases) > 1:
+                    self.reconcile_cases(referral_cases, occurrence_case_id)
+        self.email_report()
```
Nice, this is a simple one.

esoergel reviewed (commented):
One of the commands seemed incomplete, is that one just intended for dry runs at this point?  This would be easier to review if it were broken up into separate PRs.

How often is this reconciliation going to be run?

mkangia committed:
extract BaseModelReconciliationCommand and use commit instead of dry_run [skip ci]

mkangia commented on custom/enikshay/management/commands/drug_resistance_reconciliation.py:
```diff
@@ -0,0 +1,233 @@
+                             "Picked first opened.")
+
+    @staticmethod
+    def _get_open_occurrence_case_ids_to_process():
```
hmm, that would do the same thing but there are couple of advantages here
1. This would currently get ids per db shard thus reducing the number
2. The list would be shorted per shard and hence the memory footprint

I could possibly try to alter the `get_open_case_ids_in_domain_by_type` but i am not confident about making that change.

Let me know if that doesn't sound a reasonable explanation for retaining this.

mkangia committed:
dry_run -> commit, use set comprehension, add case id to debug [skip ci]

mkangia committed:
fetch episode cases for all occurrence at once [skip ci]

mkangia committed:
allow drug resistance reconciliation to commit when passed since its confirmed now.
Skip logging missing person cases since those should be orphaned cases from migration clean up [skip ci]

mkangia committed:
remove redundant method [skip ci]

mkangia committed:
remove redundant if [skip ci]

mkangia committed:
move DOMAIN to base command itself [skip ci]

mkangia committed:
extract get_first_opened_case and log progress/debug details only when required [skip ci]

mkangia committed:
extra some const and fix issue when fetching first opened case [skip ci]

mkangia committed:
check the correct set of cases for count [skip ci]

mkangia committed:
update reconciliation logic under comments [skip ci]

mkangia committed:
extract get_relevant_episode_case to DRY, simply get_relevant_episode_case_to_retain [skip ci]

mkangia committed:
better naming, log if no open active episode case found for occurrence reconciliation [skip ci]

esoergel commented on custom/enikshay/management/commands/investigations_reconciliation.py:
```diff
@@ -184,26 +184,24 @@ def _get_open_person_case_ids_to_process():
+    return [case.case_id for case in all_cases if case.type == CASE_TYPE_OCCURRENCE]
 
 
 def get_open_confirmed_drtb_episode_cases(person_case_id):
```
There's an existing function `get_all_episode_cases_from_person` which does most of the work of these two functions (although it has the same issue you resolved in this PR, of not making use of a list of arguments).

esoergel commented on custom/enikshay/management/commands/base_model_reconciliation.py:
```diff
@@ -30,9 +30,7 @@ def handle(self, *args, **options):
     def public_app_case(self, person_case):
-        if person_case.get_case_property(ENROLLED_IN_PRIVATE) == 'true':
-            return False
-        return True
```
👍

esoergel commented on custom/enikshay/management/commands/base_model_reconciliation.py:
```diff
@@ -61,3 +63,7 @@ def writerow(self, row):
+
+    @staticmethod
+    def get_first_opened_case(all_cases):
+        return sorted(all_cases, key=lambda x: x.opened_on)[0]
```
nit: `min` would be more appropriate (faster too, although I doubt that's relevant here)
```python
return min(all_cases, key=lambda x: x.opened_on)
```

esoergel commented on custom/enikshay/management/commands/drug_resistance_reconciliation.py:
```diff
@@ -110,34 +113,33 @@ def reconcile_drug_resistance_cases(self, drug_resistance_cases, drug_id, occurr
         # keep the one opened first, close rest.
         else:
-            retain_case = sorted(drug_resistance_cases, key=lambda x: x.opened_on)[0]
+            retain_case = sorted(drug_resistance_cases)
```
shouldn't this use `get_first_opened_case`?

esoergel replied:
nevermind, looks like you already fixed it.

mkangia committed:
update logic to retain case in case multiple open referrals [skip ci]

mkangia committed:
Merge master: Resolve conflict in custom/enikshay/tasks.py

esoergel reviewed (approved):
Is this already being used and QA'd?  The changes that keep coming in seem like iterations on something being tested.  It would be nice to merge this (and comment out the task if needed until that's ready), so that future updates can be small, and we don't need to worry about merge conflicts and the costs of long-running branches.

mkangia commented:
thank you @esoergel 

Ya merging it and then would make any small improvements as follow up.

There are total 4 reconciliations here:
1. 2 of them are ready and tasted, duplicate occurrence & episode + drug resistance reconciliation 
2. 1 is ready but not fully tested, duplicate referral
3. 1 is not ready but still kept around to find any cases that would need reconciliation while we figure out the reconciliation to be done.
